src/crafting.cpp

```diff
@@ -922,43 +922,60 @@ bool Character::craft_proficiency_gain( const item &craft, const time_duration &
         cata::optional<time_duration> max_experience;
     };
 
-    // The proficiency, and the multiplier on the time we learn it for
-    std::vector<learn_subject> subjects;
-    for( const recipe_proficiency &prof : making.proficiencies ) {
-        if( !_proficiencies->has_learned( prof.id ) &&
-            prof.id->can_learn() &&
-            _proficiencies->has_prereqs( prof.id ) ) {
-            learn_subject subject{
-                prof.id,
-                prof.learning_time_mult / prof.time_multiplier,
-                prof.max_experience
-            };
-            subjects.push_back( subject );
+    const std::vector<Character *> all_crafters = [&]() {
```

Comment:
Why is this a lambda? It obscures the meaning and makes debugging more complex.

Reply:
To keep the vector const despite non-const initialization. https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-lambda-init explains it better than I could.

I can inline it, but then it would be mutable.

Reply:
Please inline it, I don't really buy the benefit of this contrived structure to limit the scope of the non-const container that is very obviously not intended to be modified from 50 lines of code to 5.
Also, this is sufficient:
```
        std::vector<Character *> all_crafters( get_crafting_helpers() );
        all_crafters.push_back( this );
```
Reserve is a performance optimization that is clearly not going to be pertinent here.

Alternately, you can hoist this logic to a helper method get_all_crafters() with the above implementation and initialize your const vector from there.

Reply:
Because there's no `vector<npc *>` -> `vector<Character *>` conversion, I've used something slightly different. I agree the reserve is unnecessary here.